lib/streamlit/server/Server.py

```diff
@@ -86,6 +89,44 @@ class State(Enum):
     STOPPED = "STOPPED"
 
 
+class ExceededRetries(Exception):
+    pass
+
+
+def server_port_is_manually_set():
+    return config.is_manually_set("server.port")
+
+
+def start_listening(app, call_count=0):
+    """makes the server start listening at the configured port.
+    In case the port is already taken it tries listening to the next available port.
+    It will error after a 100 attempts."""
+    port = config.get_option("server.port")
+    try:
+        app.listen(port)
+    except OSError as e:
+        if e.errno == errno.EADDRINUSE:
+            if call_count >= 100:
+                raise ExceededRetries(
+                    "Port %s already in use, were unable to find a free port after a hundred attempts",
+                    port,
+                )
+            if server_port_is_manually_set():
+                LOGGER.debug("Port %s already in use, trying to use the next one", port)
+                port += 1
+                # save port 3000 because it is used for the development server in the front end
+                if port == 3000:
+                    port += 1
+
+                config._set_option("server.port", port, "server initialization")
+                start_listening(app, call_count + 1)
+            else:
+                LOGGER.error("Port %s is already in use", port)
+                sys.exit(signal.NSIG)
```

Comment:
Why use NSIG here? I may be wrong but I think signals are unrelated to exit codes...

Maybe just use sys.exit(1) ?

Reply:
I wasn't sure either but it turns out that signals are just integers. Changed it to 1.
<img width="163" alt="Screen Shot 2019-09-26 at 11 18 57 AM" src="https://user-images.githubusercontent.com/934511/65696409-8894c480-e04f-11e9-83f4-21cae1700fc7.png">

Reply:
in fact no, some are classes and some are just int.
<img width="206" alt="Screen Shot 2019-09-26 at 12 13 53 PM" src="https://user-images.githubusercontent.com/934511/65701134-2a6bdf80-e057-11e9-871a-9ab983443235.png">